### PR TITLE
4486 email pray and pay

### PR DESCRIPTION
**Pull request by v-anne**
This PR is to add email alerts to #1346, the pray and pay project. It follows up on #4386, which implemented much of the logic for the backend in terms of the `Prayer` model.

The logic for the email alerts is as follows:
When a document is confirmed to be available on CourtListener, the backend will identify whether any users had requested purchase of that document, and if so, will send all of them an email together informing them of its availability using BCC. The `Prayer` model will be updated to confirm that the document is now available, which will remove the document from the publicly available display of open requests.

### CODE REVIEW

**Comment by v-anne** [comment]
@mlissner @albertisfu

Some progress has been made on the email functionality. I looked at `cl/alerts/management/commands/cl_send_alerts.py` to see how emails are implemented in other parts of CourtListener. I do have some questions:
- Could you confirm that I have correctly implemented how to extract the url link to the file?
- Should I include text similar to the following from `alert_email.txt`?
```

 - View this item on our site: https://www.courtlistener.com{{result.absolute_url}}
{% if result.download_url %} - Download original from the court: {{result.download_url}}
{% endif %}{% if result.local_path %} - Download the original from our backup: https://storage.courtlistener.com/{{ result.local_path }}{% endif %}{% endfor %}
```

**Comment by mlissner** [comment]
Everything looks basically right to me, but I'll let Alberto take a more careful look.

**Comment by albertisfu** [comment]
@v-anne I've applied some tweaks to the PR so that email sending works properly.

One of the changes is to use email bulk sending, allowing us to send one email per recipient instead of using BCC. I think this is better due to AWS restrictions of 50 recipients when using the BCC or CC fields:
https://docs.aws.amazon.com/ses/latest/dg/quotas.html

This way, even if we have more than 50 users who need to be notified after a prayer is granted, all of them can receive the message.

- I also added tests to confirm that email sending works properly.

@mlissner I think the only thing left to confirm is whether the email template is complete or if it requires some tweaks. Currently, the emails look as follows:

Text version:

```
CourtListener.com
*****************
-------------------------------------------------------
We have news regarding your alerts at CourtListener.com
-------------------------------------------------------

A document you requested to be purchased is now available on CourtListener.
-------------------------------------------------------

 - View this item on our site: https://www.courtlistener.com/docket/121/1/hoffman-v-mason-yu/


************************
This alert brought to you by the 501(c)(3) non-profit Free Law Project
 - Blog: https://free.law
 - BlueSky: https://bsky.app/profile/free.law
 - X: https://x.com/freelawproject
 - Donate: https://donate.free.law/forms/supportflp
 - Become a Member: https://donate.free.law/forms/membership

Please donate to support our work.
```

HTML version:
Screenshot 2024-09-27 at 3 04 19 p m

**Comment by mlissner** [comment]
I made a little template that cribs from the alert email: 

https://docs.google.com/document/d/1jGjdZPueFXRHILqgI3Pp606EzrEaxvNTkcLdWNJdHTs/edit?usp=sharing

I think it should be pretty easy, but some parts of it might be tricky and it probably isn't worth spending a ton of time on all of it. So consider it a notion, but hopefully one that gives some direction! 

Thank you both!

**Comment by v-anne** [comment]
@mlissner, I left some comments on your Google doc.

**Comment by v-anne** [comment]
I added three things as you suggested in the Google Doc:
- price of the document
- date requested
- how many people were waiting for it.

Despite our initial hesitance about the date requested component, I decided to include it. I think limiting the query for dates to the QuerySet described below makes this relatively efficient and is a minor use of resources. If you disagree, I'm happy to remove it.

```python3
open_prayers = Prayer.objects.filter(
            recap_document=instance, status=Prayer.WAITING
        ).select_related("user")
```

**Comment by v-anne** [comment]
@mlissner @albertisfu

I attempted adding the case name and court number to the alerts. I think the case name works but the court name doesn't. I couldn't quite figure out how `{{ result.court_citation_string|nbsp|safe }}` works in the `alert_email.html` file.

Otherwise, I think this is almost ready to merge, after one more review.

**Comment by mlissner** [comment]
You just have to follow the FK chain. So if you have a RECAPDocument, the full chain would be:

`rd.docket_entry.docket.court.whatever_field`

I think the correct field on Court is `full_name`, but I'm not sure. Want to try that?

**Comment by albertisfu** [comment]
I've tweaked the templates and email context to make them look more like the email template Mike shared.

Added additional test assertions.

Email subject remains as: `A document you requested is now on CourtListener `

Screenshot 2024-09-30 at 11 58 33 a m


```
*****************
CourtListener.com
*****************
-------------------------------------------------------
Your Wish Has Been Granted!
-------------------------------------------------------

The document you were waiting for is now available in RECAP.
-------------------------------------------------------

Velazquez-Watts v. Roberts (4:59-ms-57576)
Thirteenth circuit of the Zoo
~~~
View Docket: https://www.courtlistener.com/docket/342/velazquez-watts-v-roberts/?order_by=desc

Document Number: 6
Date Filed: Aug 16, 2015
Description: Dismissing Case
View Document: https://www.courtlistener.com/docket/342/1/velazquez-watts-v-roberts/
~~~
You requested it on Sep 30, 2024
2 people were also waiting for it.
Somebody paid $1.00 to make it available to all of us.

************************
This alert brought to you by the 501(c)(3) non-profit Free Law Project
 - Blog: https://free.law
 - BlueSky: https://bsky.app/profile/free.law
 - X: https://x.com/freelawproject
 - Donate: https://donate.free.law/forms/supportflp
 - Become a Member: https://donate.free.law/forms/membership
```

**Comment by v-anne** [comment]
I think this is good to go. Thank you for your work on this, Alberto!

**Review by mlissner — COMMENTED** [review]
One quick comment from me, but Alberto, I leave it to you for full review.

**Comment by mlissner** [comment]
Thank you @v-anne for sticking with this big feature!

File: cl/search/signals.py
Comment (albertisfu) [reply cl/search/signals.py:580]: I agree with Mike. We could move this block of code to `favorites.utils` and simply call the method from here.

**Review by albertisfu — COMMENTED** [review]
In addition to Mike's suggestion, I would also recommend moving `prayer_email.html` and `prayer_email.txt` to `favorites/templates`.

With those changes, I think this is good to go.

Thanks @v-anne

File: cl/search/signals.py
Comment (v-anne) [reply cl/search/signals.py:580]: Ok, I'll make this change.

**Comment by v-anne** [comment]
Both of those comments have now been resolved.

**Comment by albertisfu** [comment]
> Both of your comments have now been resolved.

Thanks LGTM!

**Review by mlissner — APPROVED** [review]
Very cool. Thank you both. 

@v-anne, deployments are blocked atm, but this will go live some time tomorrow, I expect. 

Onward to the last pieces!

**Pull request merged**
